### DISCUSSION TITLE
Simple question, changing list elements

### DISCUSSION THREAD

**Discussion by nathanieltagg** [cue-lang/cue#1356] [Q&A] [answered]
I'm just looking at CUE as a beginner, wondering if it's suitable for a project.

I'm confused about specializing list elements.  A simple example:

```
out: [ {a:int},
       {a:3}
       ]
       
out[0]:a:2

```
That second statement fails. I can't figure out the right syntax to get
```
out: [ {a:2}, {a:3} ]
```

**Comment by eonpatapon** [comment]
You need to redefine the list with your concrete values at the correct index so that unification works:

```cue
out: [{a: int}, {a: 3}]
out: [{a: 2}, _]
```

**Comment by eadlam** [comment]
My approach is to never work with lists, but rather generate them as needed. For example:

```
_data: {
    "0": {a:int}
    "1": {a:int}
    "2": {a:int}
    "3": {a:int}
    "4": {a:int}
}

_data: "2": {a: 3}

out: [for k, v in _data {v}]
```
 Results in:
```
$ cue eval list.cue 
out: [{
    a: int
}, {
    a: int
}, {
    a: 3
}, {
    a: int
}, {
    a: int
}]
```

**Reply by eadlam** [reply]
Alternatively, if the keys in the objects are unique, you can index by that:

```
out: [
    {a:int},
    {b:int},
    {c:int},
    {d:int},
    {e:int}
]

out: [...{[Name="c"]: 3}]
```

**Reply by verdverm** [reply]
note, you do not need the `Name=` in there

`out: [...{["c"]: 3}]` is sufficient for the same effect

Neat pattern though!

**Comment by nathanieltagg** [comment]
Assuming that I don’t go with the named properties, is it possible to address the nth element?

The extra translation stage seems a bit arbitrary and unreadable.

**Reply by verdverm** [reply]
Generally you cannot "set" a specific element by index. Recall the CUE is going to unify the various statements you write together. So when you write lists like

```cue
// a list of two ints
l: [ int, int ]
// a list containing any and 2 ("set" the second element)
l: [ _, 2 ]
// a list containing 1 and any ("set" the first element)
l: [ 1, _ ]
```

they are implicitly being unified. This is like @eonpatapon 's response. You will need to ensure that all statements have the same length for the list

You could do something with list comprehension, but this implies making a new list at a new label.

I personally tend to start with a list, convert to a struct when doing things, and then convert back to a list. Something like what is here: https://cuetorials.com/patterns/unique-list/

**Comment by myitcv** [comment, marked as answer]
@nathanieltagg welcome to the CUE project! A number of good answers already, but I just wanted to add that your question is one of the motivations for #165. Taking your original example, the following would be allowed under that proposal:

```
out: 0: a: 2
```
